File: extras/ghidra_plugin/src/main/java/sda/util/ObjectHash.java
package sda.util;

public class ObjectHash
{
    private String hashContent = "";
    private Long hash = 0L;

    public ObjectHash() {
    }

    public ObjectHash(Long hash)
    {
        this.hash = hash;
    }

    public ObjectHash(Long hash, String hashContent)
    {
        this.hash = hash;
        this.hashContent = hashContent;
    }

    public void addValue(String value) {
        hashContent += "{" + value + "}";
    }

    public void addValue(int value) {
        addValue((long)value);
    }

    public void addValue(Long value) {
        addValue(value.toString());
    }

    public long getHash() {
        return hash*31 + hash(hashContent);
    }

    public void join(ObjectHash hash) {
        this.hash = this.hash * 31 + hash.getHash();
    }
    public void add(ObjectHash hash) {
        this.hash = this.hash + hash.getHash();
    }

    private static long hash(String string) {
        long h = 1125899906842597L;
        for (int i = 0; i < string.length(); i++) {
            h = 31*h + string.charAt(i);
        }
        return h;
    }
}

File: src/utilities/ObjectHash.cpp
#include "ObjectHash.h"

void ObjectHash::addValue(std::string value) {
	m_hashContent += "{" + value + "}";
}

void ObjectHash::addValue(int value) {
	addValue((int64_t)value);
}

void ObjectHash::addValue(int64_t value) {
	addValue(std::to_string(value));
}

ObjectHash::Hash ObjectHash::getHash() {
	return m_hash * 31 + hash(m_hashContent);
}

void ObjectHash::join(ObjectHash& hash) {
	m_hash = m_hash * 31 + hash.getHash();
}

void ObjectHash::add(ObjectHash& hash) {
	m_hash = m_hash + hash.getHash();
}

ObjectHash::Hash ObjectHash::hash(std::string string) {
	Hash h = 1125899906842597L;
	for (int i = 0; i < string.length(); i++) {
		h = 31 * h + string.at(i);
	}
	return h;
}
